### DISCUSSION TITLE
Know which feature is applied in the hb-shape() trace

### DISCUSSION THREAD

**Discussion by murasu** [harfbuzz/harfbuzz#3014] [Q&A] [answered]
This is perhaps related to #2962 

The GSUB part of the trace output gives the lookup numbers like:
```
trace: start lookup 14	buffer: ...
trace: end lookup 14	buffer: ...
```
Will it be possible to also get the GSUB feature that is being applied?

Is this https://github.com/justvanrossum/fontgoggles/pull/62 addressing that?

**Comment by jfkthame** [comment]
It seems like this may be a tricky question to answer in the general case, because a given lookup could be referenced by multiple different features.

**Reply by murasu** [reply]
Yes, I can understand that. Will it be possible to have something like:
```
trace: start feature pref buffer ...
... all other lookups ...
trace: end feature pref buffer ...
```
Perhaps with a flag that requests for deeper traces?

**Comment by behdad** [comment, marked as answer]
This keeps coming up. We currently don't keep that information around. Is up to a debugger to connect back lookups to features. But I understand that the trace by itself is so close to all one needs, that adding this feature will make it almost needless of post-processing.

**Reply by behdad** [reply]
Let me think about it.

**Reply by murasu** [reply]
Thank you!

**Comment by khaledhosny** [comment]
While this is currently tricky, I think @simoncozens’s https://github.com/simoncozens/crowbar tries to guess the feature, and if the font is built with Debg (?) table (e.g. with fontTools feaLib) it can even give lookup names and line numbers.
